swift/extractor/remapping/SwiftFileInterception.cpp

```diff
@@ -0,0 +1,169 @@
+#define FILE_CREATION_MODE (O_CREAT | O_TMPFILE)
+#endif
+
+namespace fs = std::filesystem;
+
+namespace {
+namespace original {
+
+void* openLibC() {
+  if (auto ret = dlopen(SHARED_LIBC, RTLD_LAZY)) {
+    return ret;
+  }
+  std::cerr << "Unable to dlopen " SHARED_LIBC "!\n";
+  std::abort();
+}
+
+void* libc() {
+  static auto ret = openLibC();
+  return ret;
+}
+
+template <typename Signature>
+Signature get(const char* name) {
+  return reinterpret_cast<Signature>(dlsym(libc(), name));
+}
+
+int open(const char* path, int flags, mode_t mode = 0) {
+  static auto original = get<int (*)(const char*, int, ...)>("open");
+  return original(path, flags, mode);
+}
+
+}  // namespace original
+
+bool endsWith(const std::string_view& s, const std::string_view& suffix) {
+  return s.size() >= suffix.size() && s.substr(s.size() - suffix.size()) == suffix;
+}
+
+auto& fileInterceptorInstance() {
+  static std::weak_ptr<codeql::FileInterceptor> ret{};
+  return ret;
+}
+
+bool mayBeRedirected(const char* path, int flags = O_RDONLY) {
+  return (!fileInterceptorInstance().expired() && (flags & O_ACCMODE) == O_RDONLY &&
+          endsWith(path, ".swiftmodule"));
+}
+
+}  // namespace
+
+namespace codeql {
+
+class FileInterceptor {
+ public:
+  FileInterceptor(fs::path&& workingDir) : workingDir{std::move(workingDir)} {
+    fs::create_directories(hashesPath());
+    fs::create_directories(storePath());
+  }
+
+  int open(const char* path, int flags, mode_t mode = 0) const {
+    fs::path fsPath{path};
+    assert((flags & O_ACCMODE) == O_RDONLY);
+    errno = 0;
+    // first, try the same path underneath the artifact store
+    if (auto ret = original::open(redirectedPath(path).c_str(), flags);
+        ret >= 0 || errno != ENOENT) {
+      return ret;
+    }
+    errno = 0;
+    // then try to use the hash map
+    if (auto hashed = hashPath(path)) {
+      if (auto ret = original::open(hashed->c_str(), flags); ret >= 0 || errno != ENOENT) {
+        return ret;
+      }
+    }
+    return original::open(path, flags, mode);
+  }
+
+  fs::path redirect(const fs::path& target) const {
+    assert(mayBeRedirected(target.c_str()));
+    auto ret = redirectedPath(target);
+    fs::create_directories(ret.parent_path());
+    if (auto hashed = hashPath(target)) {
+      std::error_code ec;
+      fs::create_symlink(ret, *hashed, ec);
+      if (ec) {
+        std::cerr << "Cannot remap file " << ret << " -> " << *hashed << ": " << ec.message()
+                  << "\n";
+      }
+    }
+    return ret;
+  }
+
+ private:
+  fs::path hashesPath() const { return workingDir / "hashes"; }
+
+  fs::path storePath() const { return workingDir / "store"; }
+
+  fs::path redirectedPath(const fs::path& target) const {
+    return storePath() / target.relative_path();
+  }
+
+  std::optional<fs::path> hashPath(const fs::path& target) const {
+    if (auto fd = original::open(target.c_str(), O_RDONLY | O_CLOEXEC); fd >= 0) {
+      return hashesPath() / hashFile(fd);
```

Comment:
as you state in a follow-up comment, this is just a mapping from hashes to file implemented in the filesystem. So if one wants to read the file corresponding to a certain hash `XXXX`, it suffices to read `hashes/XXXX`, as it will be a symlink to the required file. For this, file moves (of the original artifact) make no difference.

Reply:
Ok. Thanks. Is the structure of the artifact directory documented somewhere?

Reply:
Minor nit: The fact that `hashFile` actually closes the given open handle breaks symmetry, and may become quite error prone. I think the code would be less error prone if `hashFile` took a path as input and returned an optional hash string.

Reply:
that wouldn't work, unless we called it `hashOriginal` or `hashReal`, because of open redirection. We could integrate that functionality in the remapping library, but it would be a bit crammed. I don't mind it closing the handle, as long as it is clearly documented.

Reply:
I can actually move the `hashFile` function to an internal implementation detail, but I would rather do it in my upcoming linkage awareness PR (or PRs), as I'm touching this part of the code any way.

Reply:
Sounds reasonable